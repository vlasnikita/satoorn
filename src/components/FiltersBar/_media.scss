@media only screen
and (max-width : 1250px){
  .FiltersBar__body{
    justify-content: center;
  }
}

@media only screen
and (max-width : 850px){
  .FiltersBar__search {
    margin-left: 10px;
  }

  .FiltersBar__block {
    margin-left: 10px;
  }

  .FiltersBar__item {
    padding: 0 10px;
  }

  .FiltersBar__searchInput {
    width: 180px;
    padding: 0 10px;
  }
}

@media only screen
and (max-width : 760px){
  .FiltersBar__body{
    padding: 0;
    margin: 0;
    justify-content: flex-start;
  }
}


@media only screen
and (max-width : 650px){

  .FiltersBar__block_toggler_filter{
    display: flex;
  }
  .FiltersBar__block_bars{
    display: none;
  }

  .FiltersBar__body_toggled{
    .FiltersBar__block_toggler_filter{
      display: none;
    }
    .FiltersBar__block_bars{
      display: flex;
    }
    .FiltersBar__block_toggler_search{
      display: flex;
    }
    .FiltersBar__search{
      display: none;
    }
  }

  .FiltersBar__item {
    padding: 0 5px;
    font-size: 1rem;
  }
}

@media only screen
and (max-width : 430px){
  .FiltersBar__searchInput {
    width: 140px;
    padding: 0 10px;
  }
}
