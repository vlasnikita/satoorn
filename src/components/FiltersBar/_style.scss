.FiltersBar{
    z-index: 1102;
    position: absolute;
    height: $headerHeight;
    top: 0;
    transition: $scrollAwayNodesTiming;
    pointer-events: none;

    &__body{
        height: 100%;
        display: flex;
        flex-direction: row;
        align-items: center;
        justify-content: flex-start;

        &_toggled{
            .FiltersBar__search{
                display: none;
            }
        }

        > * {
            pointer-events: all;
        }
    }

    &__block{
        display: flex;
        flex-direction: row;
        justify-content: flex-start;
        align-items: center;
        margin-left: 15px;
        height: 36px;
        background: #e6e6f0;
        border-radius: 6px;
        width: auto;
        box-shadow: $filtersInsetBoxShadow;

        &_toggler{
            display: none;
            padding: 0;
            margin: 0;

            &_search{
                margin-left: 0 !important;
            }
        }

        &_bars{
            border: 1px solid #ddd;
        }
    }

    &__item {
        font-size: 1.4rem;
        color: #999;
        height: 30px;
        line-height: 30px;
        padding: 0 15px;
        margin: 0 3px;
        cursor: pointer;
        border-radius: 4px;

        &:hover{
            color: $mainPurple;
        }

        &_active{
            background: #fff;
            color: $mainPurple;

            &:hover{ color: $mainPurple; }
        }
    }

    &__search{
        margin-left: 35px;
        height: 36px;
        background: #e6e6f0;
        border-radius: 6px;
        width: auto;
        display: flex;
        flex-direction: row;
        align-items: center;
        border: 1px solid #ddd;
        box-shadow: $filtersInsetBoxShadow;

        &Input{
            width: 240px;
            outline: none;
            border: none;
            font-size: 1.4rem;
            color: #222;
            height: 100%;
            padding: 0 15px;
            background: transparent;
        }

        &Button{
            background: #fff;
            height: 30px;
            width: 30px;
            min-width: 30px;
            cursor: pointer;
            outline: none;
            border-radius: 4px;
            display: flex;
            flex-direction: row;
            align-items: center;
            justify-content: center;
            margin: 0 3px;
            border: none;
        }
    }
}

.FiltersBar__input{
    width: 100%;
    outline: none;
    border: none;
    padding: 0 10px;
    padding-bottom: 3px;
    font-size: 1.4rem;
    transition: .2s;
    background: #222;
    color: #eee;
    height: calc(100% - 3px);
}

.FiltersBar__input:focus{
    padding-left: 7px;
    background: #333333;
}

.FiltersBar__go {
    background: $mainYellow;
    border: none;
    height: 26px;
    width: 26px;
    min-width: 26px;
    cursor: pointer;
    outline: none;
    border-radius: 3px;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    margin-right: 1px;
}

@import '/media';