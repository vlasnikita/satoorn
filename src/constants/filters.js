export const TEMPLATE_FILTERS = [
    {
        key: 'gaming',
        value: 'Игровые'
    },
    {
        key: 'other',
        value: 'Неигровые'
    }
]

export const CHALLENGE_FILTERS = [
    {
        key: 'created',
        value: 'Доступные'
    },
    {
        key: 'accepted',
        value: 'Принятые'
    }
]
