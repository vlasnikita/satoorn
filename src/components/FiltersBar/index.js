import React, { Component } from 'react';
import {connect} from "react-redux";

import {
    setChallengesFilter, setChallengesSearchTerm,
    setChallengesByUserIdSearchTerm, setChallengesByUserIdFilter,
    resetFilters
} from "Actions/challenges";
import { getStreamers, setStreamersSearchTerm } from "Actions/streamers";
import { CHALLENGE_FILTERS } from 'Constants/filters'

import iconLens from 'Static/lens.svg'
import filterIcon from 'Static/filter.svg'

let timer = null

class FiltersBar extends Component {

    constructor(props) {
        super(props)
        this.domNode = React.createRef();
        this.state = {
            searchTerm: '',
            isFilterTogglerActive: false
        }
    }

    componentWillUnmount(){
        this.props.handleResetFilters()
    }

    getBars = () => {
        const {
            filter, searchTerm,
            activeStreamer, activeStreamerFilter, activeStreamerSearchTerm
        } = this.props

        const filterNames = CHALLENGE_FILTERS
        const activeKey = activeStreamer.id ? activeStreamerFilter : filter
        return filterNames.map((filter, i) => (
            <div
                key={i}
                id={`FiltersBar__item_${i}`}
                className={`FiltersBar__item ${filter.key === activeKey && 'FiltersBar__item_active'}`}
                onClick={() => this.handleFilterChange(filter)}
            >{filter.value}</div>
        ))
    }

    getPlaceholder = () => {
        let placeholder
        if(this.props.isStreamer) placeholder = 'Поиск по нику'
        else placeholder = 'Поиск челленджа'

        return placeholder
    }

    handleFilterChange = filter => {
        // Фильтрация:
        // – челленджей стримера
        if(this.props.activeStreamer.id) this.props.handleSetChallengesByUserIdFilter(filter)
        // – всех челленджей на главной
        else this.props.handleSetChallengesFilter(filter)
    }

    handleSearchTermChange = e => {
        clearTimeout(timer)
        const searchTerm  = e.target.value

        this.setState({ searchTerm })

        if(this.props.isStreamer) {
            timer = setTimeout(() => {
                this.props.handleSetStreamersSearchTerm(searchTerm)
            }, 250)
        }
        else {
            timer = setTimeout(() => {
                if(this.props.activeStreamer.id) this.props.handleSetChallengesByUserIdSearchTerm(searchTerm)
                else this.props.handleSetChallengesSearchTerm(searchTerm)
            }, 250)
        }
    }

    render() {
        if(this.props.isWidgets) return (<div className='FiltersBar' ref={this.domNode}><WidgetsBar {...this.props} /></div>)
        else return (
            <div className='Container FiltersBar' ref={this.domNode}>
                <div className={`Container__body FiltersBar__body ${this.state.isFilterTogglerActive ? 'FiltersBar__body_toggled' : ''}`}>
                    <button
                        className={`FiltersBar__block FiltersBar__block_toggler FiltersBar__block_toggler_search`}
                        onClick={()=>this.setState({ isFilterTogglerActive: false })}
                    >
                        <div className="FiltersBar__searchButton">
                            <img src={iconLens} width={16}/>
                        </div>
                    </button>
                    <div className="FiltersBar__search">
                        <input
                            type="text"
                            className="FiltersBar__searchInput"
                            placeholder={this.getPlaceholder()}
                            value={this.state.searchTerm}
                            onChange={this.handleSearchTermChange}
                        />
                        <div className="FiltersBar__searchButton">
                            <img src={iconLens} alt="Поиск" width={16}/>
                        </div>
                    </div>

                    <button
                        className={`FiltersBar__block FiltersBar__block_toggler FiltersBar__block_toggler_filter`}
                        onClick={()=>this.setState({ isFilterTogglerActive: true })}
                    >
                        <div className="FiltersBar__searchButton">
                            <img src={filterIcon} width={16}/>
                        </div>
                    </button>
                    {!this.props.isStreamer && <div className="FiltersBar__block FiltersBar__block_bars">
                        {this.getBars()}
                    </div>}
                </div>
            </div>
        )
    }
}
const mapStateToProps = (state) => ({
    profile: state.profile.profile,

    filter: state.challenges.filter.key,
    searchTerm: state.challenges.searchTerm,

    activeStreamer: state.activeStreamer.activeStreamer,
    activeStreamerSearchTerm: state.challenges.activeStreamerSearchTerm,
    activeStreamerFilter: state.challenges.activeStreamerFilter.key,
});

const mapDispatchToProps = (dispatch) => {
    return {
        handleSetChallengesFilter: filter => dispatch(setChallengesFilter(filter)),
        handleSetChallengesByUserIdFilter: filter => dispatch(setChallengesByUserIdFilter(filter)),
        handleSetChallengesSearchTerm: searchTerm => dispatch(setChallengesSearchTerm(searchTerm)),
        handleSetChallengesByUserIdSearchTerm: searchTerm => dispatch(setChallengesByUserIdSearchTerm(searchTerm)),

        handleSetStreamersSearchTerm: searchTerm => dispatch(setStreamersSearchTerm(searchTerm)),
        handleGetStreamers: searchTerm => dispatch(getStreamers(searchTerm)),

        handleResetFilters: () => dispatch(resetFilters())
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(FiltersBar);
export const FiltersBarPure = FiltersBar