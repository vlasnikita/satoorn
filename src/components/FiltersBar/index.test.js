import React from 'react'
import 'jest-specific-snapshot'
import renderer from 'react-test-renderer'
import configureMockStore from "redux-mock-store";
import { Provider } from "react-redux";
import { BrowserRouter } from 'react-router-dom';

import FiltersBar from './index'
import { FiltersBarPure } from './index'
import {unmountComponentAtNode} from "react-dom";
import Adapter from 'enzyme-adapter-react-16'
import {configure, mount, shallow} from "enzyme";
import {CHALLENGE_FILTERS} from "Constants/filters";

configure({ adapter: new Adapter() })

const mockStore = configureMockStore();
const initialState = {
    challenges: {
        challenges: require('Mocks/challenges'),
        filter: CHALLENGE_FILTERS[0],
        activeStreamerChallenges: require('Mocks/challenges'),
        activeStreamerFilter: CHALLENGE_FILTERS[0],
        searchTerm: '',
        loading: false,
        loaded: false
    },
    activeStreamer: {
        activeStreamer: {
            "id": "1",
            "streamerLogin": "user_1",
            "streamerPhotoLink": "https://images-cdn.9gag.com/photo/anbZVN0_700b.jpg",
            "imageId": "overwatch_stub_id",
            "title": "some title",
            "description": "Some default description Some default description Some default description Some default description Some default description Some default description Some default description Some default description Some default description Some default description ",
            "donationAmount": 78,
            "commissionAmount": 50,
            "donatedPersonCount": 10,
            "viewCount": 200,
            "state": "CREATED",
            "startedAt": "14-07-2020 21:00 MSK",
            "remainedDaysCount": 6,
            "streamerMessage": null
        },
        loading: false,
        loaded: false,
    },
    profile: {
        profile: {
            id: '5f108c57399c8300013dad42',
            authenticationSource: 'twitch',
            externalId: '204865822',
            externalProfileLink: null,
            externalPhotoLink: 'https://static-cdn.jtvnw.net/user-default-pictures-uv/75305d54-c7cc-40d1-bb9c-91fbe85943c7-profile_image-300x300.png',
            email: null,
            login: 'vlasnikita',
            bio: null,
            walletAmount: 0,
            viewerCount: 0,
            followerCount: 0,
            challengeCount: 0,
            streamer: false,
            contacts: []
        },
        loaded: true
    }
}

const store = mockStore(initialState);
let container = null

beforeEach(() => {
    // подготавливаем DOM-элемент, куда будем рендерить
    container = document.createElement("div");
    document.body.appendChild(container);
});
afterEach(() => {
    // подчищаем после завершения
    unmountComponentAtNode(container);
    container.remove();
    container = null;
});

it('renders correctly FiltersBar', () => {
    function createNodeMock(element) {
        if (element.className === 'Challenge__description') {
            // This is your fake DOM node for <p>.
            // Feel free to add any stub methods, e.g. focus() or any
            // other methods necessary to prevent crashes in your components.
            return {};
        }
        // You can return any object from this method for any type of DOM component.
        // React will use it as a ref instead of a DOM node when snapshot testing.
        return element
    }

    const options = {createNodeMock}
    const tree = renderer.create(
        <BrowserRouter>
            <Provider store={store}>
                <FiltersBar
                />
            </Provider>
        </BrowserRouter>,
        options
    )

    expect(tree.toJSON()).toMatchSnapshot();
});