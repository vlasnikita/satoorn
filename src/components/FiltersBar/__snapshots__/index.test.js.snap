// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly FiltersBar 1`] = `
<div
  className="Container FiltersBar"
>
  <div
    className="Container__body FiltersBar__body "
  >
    <button
      className="FiltersBar__block FiltersBar__block_toggler FiltersBar__block_toggler_search"
      onClick={[Function]}
    >
      <div
        className="FiltersBar__searchButton"
      >
        <img
          src="lens.svg"
          width={16}
        />
      </div>
    </button>
    <div
      className="FiltersBar__search"
    >
      <input
        className="FiltersBar__searchInput"
        onChange={[Function]}
        placeholder="Поиск челленджа"
        type="text"
        value=""
      />
      <div
        className="FiltersBar__searchButton"
      >
        <img
          alt="Поиск"
          src="lens.svg"
          width={16}
        />
      </div>
    </div>
    <button
      className="FiltersBar__block FiltersBar__block_toggler FiltersBar__block_toggler_filter"
      onClick={[Function]}
    >
      <div
        className="FiltersBar__searchButton"
      >
        <img
          src="filter.svg"
          width={16}
        />
      </div>
    </button>
    <div
      className="FiltersBar__block FiltersBar__block_bars"
    >
      <div
        className="FiltersBar__item FiltersBar__item_active"
        id="FiltersBar__item_0"
        onClick={[Function]}
      >
        Доступные
      </div>
      <div
        className="FiltersBar__item false"
        id="FiltersBar__item_1"
        onClick={[Function]}
      >
        Принятые
      </div>
    </div>
  </div>
</div>
`;
